drivers:platform:maxim Maxim platform drivers

CiprianRegus opened the pull request:
Initial implementation of peripheral drivers for MAX32660 (UART, RTC,
SPI, GPIO and IRQ).

Signed-off-by: Ciprian Regus <ciprian.regus@analog.com>

mchindri commented:
Please separate the drivers in different commits

CiprianRegus commented:
V1:
- Created individual commits for each driver.

mchindri commented on drivers/platform/maxim/maxim_uart.c:
```diff
@@ -0,0 +1,353 @@
+
+	uint32_t err = 0;
+
+	int32_t len = bytes_number;
```
I assume you have a maximum length to transfer with UART_Write, if so, I would split the transfer in multiple transfer.

CiprianRegus replied:
UART_Write handles that. It works by adding each byte to the tx fifo, and also waiting (in a loop) if the fifo is full. So in this case I don't think I have to split into multiple transfers.

mchindri commented on drivers/platform/maxim/maxim_uart.h:
```diff
@@ -0,0 +1,74 @@
+	uart_flow_pol_t pol;
+};
+
+int32_t uart_register_callback(uint8_t, struct callback_desc *);
```
Even though I would put this function here to. Callback should go in irq and use irq_register

CiprianRegus replied:
Will do that.

mchindri commented on drivers/platform/maxim/maxim_uart.c:
```diff
@@ -0,0 +1,353 @@
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+void UART0_IRQHandler()
```
Can you read some register to see which UART triggerd the function, instead of having 2 function doing the same thing?

buha replied:
Yeah valid point from Mihail above.
I think these UART0_IRQHandler, UART1_IRQHandler etc. are hardcoded with that name and the functions need to be defined. What you could do is find a Maxim HAL define that tells you if UART0 is available and do something like:
```
#ifdef MAXIM_UART0_EXISTS # <--- arbitrary name
void UART0_IRQHandler() {
    uart_irq(0);
}
#endif
#ifdef MAXIM_UART1_EXISTS
... etc.
```
where uart_irq() is a static function that will contain the same code for all uart instances.

amiclaus reviewed (commented):
there are some valid issues detected by the CI, they need to be addressed.

amiclaus commented on drivers/platform/maxim/maxim_stdio.c:
```diff
@@ -0,0 +1,134 @@
+/**
+ * @file	stdio.c
```
`@file maxim_stdio.c`

amiclaus commented on drivers/platform/maxim/maxim_rtc.c:
```diff
@@ -0,0 +1,315 @@
+ * @param tmr_cnt - New value of the timer counter.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t rtc_set_cnt(struct rtc_desc *desc, uint32_t tmr_cnt)
```
let's keep consistent, and use `dev` naming as in other functions from this file

amiclaus commented on drivers/platform/maxim/gpio_extra.h:
```diff
@@ -0,0 +1,72 @@
+int32_t max_gpio_irq_set_trigger_level(struct gpio_desc *, enum irq_trig_level);
+
+#endif
+
```
nit: extra line

amiclaus commented on drivers/platform/maxim/maxim_stdio.h:
```diff
@@ -0,0 +1,55 @@
+ *   @brief  Header file for UART driver stdout/stdin redirection.
+ *   @author Ciprian Regus (ciprian.regus@analog.com)
+********************************************************************************
+ * Copyright 2021(c) Analog Devices, Inc.
```
nit: `2022`

amiclaus commented on drivers/platform/maxim/maxim_stdio.h:
```diff
@@ -0,0 +1,55 @@
+int _fstat(int, struct stat *);
+
+#endif
+
```
nit: extra line

amiclaus commented on drivers/platform/maxim/spi_extra.h:
```diff
@@ -0,0 +1,49 @@
+extern const struct spi_platform_ops max_spi_ops;
+
+#endif
+
```
nit: extra line

buha commented on drivers/platform/maxim/maxim_uart.h:
```diff
@@ -0,0 +1,74 @@
+ */
+struct maxim_uart_init_param {
+	uint32_t port;
+	uart_parity_t parity;
```
size, parity and stop configurations already exist in uart.h - they are general for all UARTs, regardless of manufacturer/implementation.
https://github.com/analogdevicesinc/no-OS/blob/master/include/no-os/uart.h#L105-L111

You should only have in maxim_uart_init_param settings that are Maxim specific only.

CiprianRegus replied:
I no longer use that struct for init and will delete it.

buha commented on drivers/platform/maxim/maxim_spi.c:
```diff
@@ -0,0 +1,179 @@
+	descriptor->bit_order = param->bit_order;
+	descriptor->platform_ops = &max_spi_ops;
+
+	ret = SPI_Init(SPI0A, descriptor->mode, param->max_speed_hz);
```
This SPI0A should be passed in to the driver from the application that initializes the driver (unless all Maxim chips have only 1 SPI interface, which I doubt).

So you should create a maxim_spi_init_param which holds a member named e.g. "interface" or "instance", which gets initialized with SPI0A, gets saved in driver desc->extra and used thereafter.

buha commented on drivers/platform/maxim/maxim_stdio.c:
```diff
@@ -0,0 +1,134 @@
+#define STDIN_FILENO	0   /**> Definition of stdin */
+#define STDOUT_FILENO   1   /**> Definition of stdout */
+#define STDERR_FILENO   2   /**> Definition of stderr */
+#define EBADF		  -1   /**> Error code for EBADf */
```
Does Maxim toolchain not defined EBADF ? I see <errno.h> is included above.

CiprianRegus replied:
I looked it up and EBADF is available in errno.h. That #define was part of the original file provided by Maxim, but I think I can delete it.

adrimbarean commented on drivers/platform/maxim/maxim_uart.c:
```diff
@@ -0,0 +1,353 @@
+
+void UART0_IRQHandler()
+{
+	if(!cb[0])
```
I think we should clear interrupt flags even if callback is not registered. This way we avoid continuously triggering interrupts and hanging the controller.

adrimbarean commented on drivers/platform/maxim/maxim_uart.c:
```diff
@@ -0,0 +1,353 @@
+	/** Clear all interrupt flags */
+	uart_regs->int_fl = 0x0;
+
+	while(reg_int) {
+		if((reg_int & 1) && uart_regs->int_en && BIT(n_int)) {
+			void *ctx = cb[0]->ctx;
+			void *config = cb[0]->config;
+			cb[0]->callback(ctx, n_int, config);
+		}
+		n_int++;
+		reg_int >>= 1;
+	}
```
This doesn't seem to work correctly. I propose:
`while (reg_int) {
  n_int = find_first_set_bit(reg_int);
  if (reg_int & (uart_regs->int_en & BIT(n_int))) {
    //call cb
  }
  reg_int >>= n_int + 1;
}`

adrimbarean commented on drivers/platform/maxim/maxim_uart.c:
```diff
@@ -0,0 +1,353 @@
+	int32_t err = 0;
+	uint32_t bytes_read = 0;
+
+	int32_t len = bytes_number;
```
If this is done to avoid warnings, I think we can do that by using a cast.

adrimbarean commented on drivers/platform/maxim/maxim_uart.c:
```diff
@@ -0,0 +1,353 @@
+
+	int32_t len = bytes_number;
+	err = UART_Read(MXC_UART_GET_UART(desc->device_id), data, len, &bytes_read);
+
```
nitpick: extra line

adrimbarean commented on drivers/platform/maxim/maxim_uart.c:
```diff
@@ -0,0 +1,353 @@
+	read_req.data = data;
+	read_req.len = bytes_number;
+	read_req.callback = NULL;
+	err = UART_ReadAsync(MXC_UART_GET_UART(desc->device_id), &read_req);
+
+	return err;
```
We can return `UART_ReadAsync` directly here.

adrimbarean commented on drivers/platform/maxim/maxim_uart.c:
```diff
@@ -0,0 +1,353 @@
+	uart_cfg_t maxim_desc;
+	sys_cfg_uart_t maxim_desc_sys;
+
+	if(!maxim_uart || !descriptor) {
```
I would do something similar to this approach, specifically to check allocation just after it's completed and to only `free` known allocated memory. This is done because `free()` [has undefined behavior on non-null, unallocated pointers](https://stackoverflow.com/questions/5308758/can-a-call-to-free-in-c-ever-fail) and may create a [bohrbug](https://en.wiktionary.org/wiki/bohrbug).

adrimbarean commented on drivers/platform/maxim/maxim_uart.c:
```diff
@@ -0,0 +1,353 @@
+
+	NVIC_ClearPendingIRQ(MXC_UART_GET_IRQ(did));
+	NVIC_DisableIRQ(MXC_UART_GET_IRQ(did));
+	NVIC_SetPriority(MXC_UART_GET_IRQ(did), 1);
```
Question: why are we setting interrupt priority of 1 here? And is there a way where we could do this in an IRQ driver? Status quo right now is to [include the IRQ driver](https://github.com/analogdevicesinc/no-OS/blob/master/drivers/platform/xilinx/uart.c#L42) in the other ones to access IRQ specific functionality.

adrimbarean commented on drivers/platform/maxim/gpio_extra.h:
```diff
@@ -0,0 +1,72 @@
+	enum irq_trig_level mode;
+};
+
+int32_t max_gpio_disable_irq(struct irq_ctrl_desc *);
```
Here as well as with UART I think we might want to use IRQ specific methods.

adrimbarean commented on drivers/platform/maxim/maxim_gpio.c:
```diff
@@ -0,0 +1,439 @@
+
+	descriptor->number = param->number;
+	descriptor->platform_ops = param->platform_ops;
+	descriptor->extra = param->extra;
```
To use this properly the user will have to know how this driver works as well as the maxim DFP driver and know the name of the `gpio_cfg_t` object the they need to place in the initialization structure. Is there a way to infer it based on the pin number or maybe a port ID? Those things should be more straight forward to use.

adrimbarean commented on drivers/platform/maxim/maxim_gpio.c:
```diff
@@ -0,0 +1,439 @@
+	maxim_extra->func = GPIO_FUNC_OUT;
+	GPIO_Config(maxim_extra);
+
+	if(value == 0)
```
nitpick: we can use `GPIO_LOW` in the if compare. Also we need a space between if and brackets e.g. `if (condition)`

adrimbarean replied:
If we have `GPIO_HIGH_Z` option in `max_gpio_set_value` should we have it here as well?

CiprianRegus commented:
V2:
- Implemented a common IRQ handler function for both UART ports.
- Removed maxim_uart_init_param from maxim_uart.h
- SPI interfaces (SPI0A and SPI1A) are initialised based on device id.
- Removed EBADF define from maxim_stdio.c
- Fixed the UART IRQ handler as suggested and updated the other IRQ handlers to use a similar approach.
- Fixed the double free in maxim_uart.c
- Removed the NVIC function calls in maxim_uart.c
- Added max_gpio_init_param in gpio_extra.h to make the gpio init easier (and not use gpio_cfg_t directly as before).
- GPIO_HIGH_Z can now be used when setting a gpio pin as output.
- Fixed doxygen.

buha commented on drivers/platform/maxim/maxim_uart.c:
```diff
@@ -0,0 +1,323 @@
+	if (port >= N_PORTS || !cb[port])
+		return -EINVAL;
+
+	free(cb[port]);
```
set the cb[port] to NULL after free, it could be accessed again
https://stackoverflow.com/questions/1025589/setting-variable-to-null-after-free

buha commented on drivers/platform/maxim/gpio_extra.h:
```diff
@@ -0,0 +1,79 @@
+#ifndef GPIO_EXTRA_H_
+#define GPIO_EXTRA_H_
+
+#include "stdbool.h"
```
include the standard libraries with the <> characters

buha commented on drivers/platform/maxim/gpio_extra.h:
```diff
@@ -0,0 +1,79 @@
+#include "no-os/irq.h"
+#include "no-os/gpio.h"
+
+#define N_PINS	14
```
So I see that you've included these here, but I think you can drop the N_PINS and N_PORTS checks because the Maxim HAL will check these. If there is a bug in Maxim HAL and it doesn't check them, and it crashes the program, then it's fine to add the checks ourselves, but I'm pretty sure that GPIO_Config() will return some meaningful error code if you pass it port -1 and pin 1234. So you simply need to translate that error code into no-os error codes and pass it back.

CiprianRegus replied:
I checked that, and GPIO_Config() doesn't check for either the port or pin number (passing an invalid port results in a null pointer being dereferenced). So, in this case, I'll keep those defines.

buha replied:
ok that's fine, it seems they have an issue in their HAL so it's fine to keep the checks, although you might have a problem then trying to keep up with all devices' available pins, you'll have to do something like:
```
#ifdef __MAX12345
#define N_PINS 14
#elif defined(__MAX54321)
#define N_PINS 16
...
```

buha commented on drivers/platform/maxim/maxim_gpio.c:
```diff
@@ -0,0 +1,451 @@
+	descriptor->number = param->number;
+	descriptor->platform_ops = param->platform_ops;
+
+	GPIO_Config(descriptor->extra);
```
check error code for all these maxim HAL calls

buha commented on drivers/platform/maxim/maxim_gpio.c:
```diff
@@ -0,0 +1,451 @@
+
+#define N_INT	14
+
+static struct callback_desc *gpio_callback[N_INT];
```
This global could be problematic when working on maxim chips with multiple ports, I think gpio_callback should account also for ports, not only pins. Otherwise, when working with GPIO1, GPIO2 etc. on top of GPIO0, these values would get overwritten.

buha commented on drivers/platform/maxim/maxim_rtc.c:
```diff
@@ -0,0 +1,315 @@
+
+	if (RTC_Init(MXC_RTC, dev->load, maxim_init_param->ms_load,
+		     &sys_cfg) != E_NO_ERROR) {
+		ret = -1;
```
-EINVAL or something like that

buha commented on drivers/platform/maxim/rtc_extra.h:
```diff
@@ -0,0 +1,81 @@
+ * @struct rtc_init_maxim
+ * @brief maxim specific init parameter
+ */
+struct rtc_init_maxim {
```
max_rtc_init ?

buha commented on drivers/platform/maxim/rtc_extra.h:
```diff
@@ -0,0 +1,81 @@
+ * @struct maxim_rtc_alarm_desc
+ * @brief maxim specific interrupt descriptor
+ */
+struct maxim_rtc_alarm_desc {
```
also here, max_rtc_alarm_desc

buha commented on drivers/platform/maxim/rtc_extra.h:
```diff
@@ -0,0 +1,81 @@
+	uint32_t irq_time;
+};
+
+int32_t rtc_get_time(uint32_t *, uint32_t *);
```
https://github.com/analogdevicesinc/no-OS/blob/master/include/no-os/rtc.h#L100
I think you should simply implement that one and not create another function.

CiprianRegus replied:
That function is no longer needed, so I'll delete it.

buha commented on drivers/platform/maxim/rtc_extra.h:
```diff
@@ -0,0 +1,81 @@
+};
+
+int32_t rtc_get_time(uint32_t *, uint32_t *);
+int32_t rtc_enable_irq(enum rtc_interrupt_id, uint32_t);
```
For these irq/callback functions, again, I think we need to discuss as a team what to do with them, I'm not sure if this is a good way to make them public to the rest of the code.

buha commented on drivers/platform/maxim/maxim_spi.c:
```diff
@@ -0,0 +1,186 @@
+	req.len = bytes_number;
+
+	if (desc->device_id == 0)
+		SPI_MasterTrans(SPI0A, &req);
```
check error codes for all HAL functions in spi.c and translate them into no-os error codes

adrimbarean commented on drivers/platform/maxim/maxim_rtc.c:
```diff
@@ -0,0 +1,315 @@
+{
+	RTC_EnableRTCE(MXC_RTC);
+
+	/** Wait for synchronization */
+	if (RTC_CheckBusy())
+		return -EBUSY;
```
nitpick: I think the comment is a bit misleading since it doesn't do any actual waiting, just returns an error if the RTC is busy.

CiprianRegus replied:
RTC_CheckBusy actually blocks the execution until the RTC's value can be read. It returns EBUSY if it waits for too long.

adrimbarean commented on drivers/platform/maxim/maxim_rtc.c:
```diff
@@ -0,0 +1,315 @@
+
+	TMR_Enable(MXC_TMR0);
+
+	if (RTC_Init(MXC_RTC, dev->load, maxim_init_param->ms_load,
```
We could use an extra rtc descriptor for maxim to store the `MXC_RTC` element. This way the driver is scalable to maxim devices with more than one RTC.

adrimbarean commented on drivers/platform/maxim/maxim_spi.c:
```diff
@@ -0,0 +1,186 @@
+	spi_req_t req;
+
+	req.bits = 8;
+	req.ssel = 0;
```
We need this to be `desc->chip_select`.

adrimbarean commented on drivers/platform/maxim/maxim_spi.c:
```diff
@@ -0,0 +1,186 @@
+	req.tx_num = 0;
+	req.rx_data = rx_buffer;
+	req.rx_num = 0;
+	req.deass = 0;
```
I think we need this to be non-zero to release the slave after transaction.

CiprianRegus commented:
V4

- UART, GPIO and RTC irq functions are now called through platform ops.
- `maxim_irq.c` now implements a NVIC controller. Each peripheral has it's own IRQ controller.
- SPI now releases the CS line after the end of transfer.
- The CS pin can now be selected using the chip_select field.
- Added some checks for SDK function calls return values. In some places those are ignored because the functions cannot return an error (the parameters are already validated in the caller).
- Included an #ifdef to guarantee that the code compiles even though the chip doesn't have an SPI1A port.
- Renamed `maxim_rtc_alarm_desc` to `max_rtc_alarm_desc`.
- `gpio_callback` is now a 2D array to account for maxim chips with multiple GPIO ports.
- Fixed uart_read_nonblocking not reading from RX fifo.
- Fixed a few minor bugs.

buha reviewed (approved):
looks good!

CiprianRegus commented:
V5:

- SPI CS is now active low.
